### PR TITLE
Setup/remove PVS proxy OVS rules on VIF plug/unplug

### PR DESCRIPTION
**Pull request by robhoes**

### CODE REVIEW

**Comment by robhoes** [comment]
This needs the final commit in xenserver/xen-api#117.

**Comment by robhoes** [comment]
@johnelse, @lindig, could you review this please?

**Comment by robhoes** [comment]
@anoobs, could you review   649c50f please?

File: scripts/setup-pvs-proxy-rules [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+if [[ $# -ne 2 ]]
+then
+    echo "Usage: $0 [add|remove] <private_xs_path>"
+    exit 1;
+fi
+
+handle_error()
+{
+    logger -t setup-pvs-proxy-rules "Failed to read $1 from xenstore"
```
Comment (lindig) [inline scripts/setup-pvs-proxy-rules]: I'd recommend to have a variable `LOG_TAG="setup-pvs-proxy-rules"` to use when calling `logger`. I'd also expect an error message on stderr, not just to syslog.

File: scripts/setup-pvs-proxy-rules [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                    ovs-ofctl --strict del-flows $PVS_BRIDGE priority=$((PVS_RULE_PRIO-1)),udp,dl_dst=$PVS_VM_MAC,nw_src=$PVS_SERVER_IP,tp_dst=$i
+                done
+            done
+            unset IFS
+        done
+        ;;
+    reset)
+        ovs-ofctl del-flows $PVS_BRIDGE
+        ovs-ofctl --strict add-flow $PVS_BRIDGE priority=0,actions=NORMAL
+        ;;
```
Comment (lindig) [inline scripts/setup-pvs-proxy-rules:133]: You want to catch `*)` and issue an error message.

**Comment by lindig** [comment]
You can get lots of good feedback by feeding shell scripts into http://www.shellcheck.net.

File: scripts/setup-pvs-proxy-rules [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+PVS_SERVER_NUM=$(xenstore-read $path)
+if [ $? -ne 0 -o -z "$PVS_SERVER_NUM" ]; then
+    handle_error $path
+fi
+
+PVS_RULE_PRIO=1000
+
+case $1 in
+    add)
+        PVS_INTERFACE_MAC=`ovs-vsctl get interface $PVS_INTERFACE mac_in_use | tr -d '"'`
```
Comment (lindig) [inline scripts/setup-pvs-proxy-rules]: `$(..)` (as used in other places) is better than using back ticks.

File: scripts/setup-pvs-proxy-rules [hunk 4]
```diff
@@ -0,0 +1,123 @@
+#!/bin/bash
+
+if [[ $# -ne 2 ]]
+then
+    echo "Usage: $0 [add|remove] <private_xs_path>"
```
Comment (lindig) [inline scripts/setup-pvs-proxy-rules]: `echo "Usage: $0 [add|remove] <private_xs_path>" >&2` directs this to stderr. This avoids that a process consuming the output would read the error message.

**Comment by robhoes** [comment]
@lindig, that website looks very handy indeed. It recommends lots of "Double quote to prevent globbing and word splitting"...

**Comment by lindig** [comment]
Yes. Any time when you can't be 100% certain that `$word` does not contain spaces or globbing characters you need to use `"$word"` or `"${word}"`. Otherwise `$word` is split into multiple words, each becoming a separate argument where you are using it.

**Comment by johnelse** [comment]
The code generally looks OK to me. I will defer to @lindig's bash knowhow though :-)

**Comment by robhoes** [comment]
@lindig I've addressed your feedback in the new commit for review (which I'll fold back into the original before merging).

**Comment by lindig** [comment]
This looks good. Shell script are notorious for subtle quoting and portability problems. There is no `reviewed` label available but I'd add it if it were.

**Comment by robhoes** [comment]
@anoobs also said he's happy :)

**Comment by anoobs** [comment]
lgtm

**Pull request merged**